Code/src/main/java/nl/utwente/viskell/ui/components/Block.java

```diff
@@ -151,46 +162,73 @@ public void handleConnectionChanges(boolean finalPhase) {
     /**
      * @return The local expression this Block represents.
      */
-    public abstract Expression getLocalExpr();
+    public abstract Pair<Expression,Set<Block>> getLocalExpr();
     
     /**
      * @return A complete expression of this block and all its dependencies.
      */
     public final Expression getFullExpr() {
-        LetExpression fullExpr = new LetExpression(this.getLocalExpr());
-        this.extendExprGraph(fullExpr);
+        Pair<Expression, Set<Block>> localExpr = getLocalExpr();
+        
+        LetExpression fullExpr = new LetExpression(localExpr.a, false);
+        Set<Block> surroundingBlocks = localExpr.b;
+        extendExprGraph(fullExpr, null, surroundingBlocks);
```

Comment:
I dislike passing a hard `null` here - either note (in the Javadoc, for example) that nulls are fine, or use `Optional`.